src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,196 @@
+                }
+                prev = bucket;
+                bucket = bucket.next;
+            }
+            prev.next = new Node<>(hash, key, value, null);
+        }
 
+        if (++size > threshold) {
+            resize();
+        }
     }
 
     @Override
     public V getValue(K key) {
+        int hash = hash(key);
+        Node<K, V> bucket = table[getBucketIndex(hash)];
+        while (bucket != null) {
+            if (
+                    bucket.hash == hash
+                    && (bucket.key == key || (key != null && key.equals(bucket.key)))
+            ) {
+                return bucket.value;
+            }
+            bucket = bucket.next;
+        }
+        return null;
+    }
+
+    @Override
+    public V remove(K key) {
+        if (!isEmpty()) {
+            int hash = hash(key);
+            int bucketIndex = getBucketIndex(hash);
+            Node<K, V> prev = null;
+            Node<K, V> bucket = table[bucketIndex];
+            while (bucket != null) {
+                if (
+                        bucket.hash == hash
+                        && (bucket.key == key || (key != null && key.equals(bucket.key)))
+                ) {
+                    if (prev == null) {
+                        table[bucketIndex] = bucket.next;
+                    } else {
+                        prev.next = bucket.next;
+                    }
+                    bucket.next = null;
+                    --size;
+                    return bucket.value;
+                }
+                prev = bucket;
+                bucket = bucket.next;
+            }
+        }
         return null;
     }
 
+    @Override
+    public List<K> getKeys() {
+        K[] keys = (K[]) new Object[size];
+        for (int i = 0, j = 0; i < table.length; i++) {
+            Node<K, V> bucket = table[i];
+            while (bucket != null) {
+                keys[j++] = bucket.key;
+                bucket = bucket.next;
+            }
+        }
+        return List.of(keys);
+    }
+
+    @Override
+    public List<V> getValues() {
+        V[] keys = (V[]) new Object[size];
+        for (int i = 0, j = 0; i < table.length; i++) {
+            Node<K, V> bucket = table[i];
+            while (bucket != null) {
+                keys[j++] = bucket.value;
+                bucket = bucket.next;
+            }
+        }
+        return List.of(keys);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private static int hash(Object key) {
+        return key == null ? 0 : key.hashCode();
+    }
+
+    private int getBucketIndex(int hash) {
+        return table.length - 1 & hash;
+    }
+
+    /**
+     * Because we are using power-of-two expansion,
+     * the elements from each bucket must either stay at same index,
+     * or move with a power of two offset in the new table.
+     */
```

Comment:
Yes, i took this logic from `java.util.HashMap` source code, because i couldn't figure out another way of redistributing Nodes within new table, and i didn't want to just add them, creating new objects, like this:
```
  table[getBucketIndex(bucket.hash)] = new Node<>(bucket.hash, bucket.key, bucket.value, null);
```